readme.md

```diff
@@ -58,6 +60,70 @@ The second (optional) argument may include the properties below.
  * `createCwd` - `boolean` Create a directory for `process.cwd()`.  This is `true` by default.
  * `createTmp` - `boolean` Create a directory for `os.tmpdir()`.  This is `true` by default.
 
+### Mapping real files & directories
+
+If you want to add files from your real filesystem to the mocked FS, there are several mapping methods to help. 
+
+#### Notes
+
+- Each mapper function duplicates all stat information (dates, permissions, etc) 
+- By default, all files are lazy-loaded, unless you specify the `{ lazyLoad: false }` option, which makes it reasonable 
+to expose large areas of the filesystem to your mocked FS.
+
+#### <a id='mappingoptions'>options</a>
+
+| Option    | Type    | Default | Description |
+| --------- | ------- | ------- | ------------
+| lazyLoad  | boolean | true    | File content isn't loaded until explicitly read
+| recursive | boolean | true    | Load all files and directories recursively (applies to `mock.mapDir()` & `mock.mapPaths()`)
+ 
+#### `mock.mapPaths(path, options)`
+
+Quickly map one or more paths. 
+
+_Note: Mapping a path makes its absolute path available in the Mock Filesystem. It is a good way to 'mount' sections of your 
+real filesystem, without allowing changes to affect the real files._
+
+```js
+// Mock FS, with single dir
+mock(mock.mapPaths('/path/to/dir'));
+
+// Mock FS, with two dirs and a file (do not recurse and pre-load all content)
+mock(mock.mapPaths([ '/path/to/dir', '/path/to/anotherdir', '/srv/some-file.txt' ], { recursive: false, lazyLoad: false }));
+```
+
+#### `mock.mapDir(dirPath, options)`
+
+Create a `Directory` for a real dir.
+
+```js
+mock({
+  // Recursively loads all node_modules
+  'node_modules': mock.mapDir(path.resolve(__dirname, '../node_modules')),
+
+  // Creates a directory named /tmp with only the files in /tmp/special_tmp_files (no subdirectories), pre-loading all content
+  '/tmp': mock.mapDir('/tmp/special_tmp_files', { recursive: false, lazyLoad:false }),
+
+  'fakefile.txt': 'content here'
+});
+```
+
+#### `mock.mapFile(filePath, options)`
+
+Create a `File` for a real file.
+
+```js
+mock({
+  // Lazy-load file
+  'my-file.txt': mock.mapFile(path.resolve(__dirname, 'assets/special-file.txt')),
+  
+  // Pre-load js file
+  'ready.js': mock.mapFile(path.resolve(__dirname, 'scripts/ready.js'), { lazyLoad: false }),
+
+  'fakefile.txt': 'content here'
+});
+```
+
```

Comment:
I appreciate all that you've put together here, @nonara.

I have some concern about the size of the change to the API (adding `bypass`, `enable`, `disable`, `mapPaths`, `mapFile`, and `mapDir`).  And I think the subtle difference in usage of `mapPaths` vs `mapFile` or `mapDir` is error prone.

It feels like we could try to minimize the changes to the API while still providing most of the same functionality.

All of `mapPaths`, `mapFile`, and `mapDir` look to be about providing a convenient way to load up a mock filesystem with data from the real filesystem.  Words like `load`, `clone`, and `read` come to mind for function names.

```js
mock({
  '/path/to/dir': mock.load('/path/to/dir'),
  '/path/to/file': mock.load('/path/to/file')
})
```

That implies adding a single `load` function that takes either a path to a directory or a path to a file and copies contents and metadata into a corresponding mock filesystem entry.  I assume we should also make it work for symlinks.

If we find that usage patterns make it very awkward to repeat the real path in the mock path, we could add a function that makes that more convenient later.

The `bypass`, `enable`, and `disable` functions feel like they could be boil down to a single function that works with either an async/promise returning function or a sync function.  In the docs below, you warn against using async calls with the mock disabled - that same warning could be included in the `bypass` docs without adding the `enable` and `disable` functions.

Having a single `bypass` function also gets around the issue that calling `enable` twice makes `disable` not work (this is fixable, but could also be avoided by not adding these functions at all).

Reply:
> All of `mapPaths`, `mapFile`, and `mapDir` look to be about providing a convenient way to load up a mock filesystem with data from the real filesystem. Words like `load`, `clone`, and `read` come to mind for function names.

I agree that `mapDir` and `mapFile` could be merged, and `load` is a great name.

As for mapPaths, this is a simple way to quickly load multiple real paths.

```ts
// With (simple)
mock(mock.loadPaths(realPaths));

// Without (simple)
mock(realPaths.reduce((p,c) => { p[c] = map.load(p) }, {})));

// With (extended)
mock{{
  ...mock.loadPaths(realPaths),
  'extra': 'content'
});

// Without (extended)
mock({
  ...realPaths.reduce((p,c) => { p[c] = map.load(p) }, {})),
  'extra': 'content'
});
```   

Bear in mind that while reduce is ugly and difficult to read, this is the shortest solution. Most people will end up using multi-line for loops, etc, all to accomplish what many have requested - a feature to simply 'mount' real areas. Worse yet, some will end up doing

```
{ 
  '/path/to/my/location': map.load('/path/to/my/location'),
  '/path/to/other.location': map.load('/path/to/other/location'),
  ...
}
```

Tedious.

The decision lies with you. I really have spent more time on this than I am able, so I'll go with whatever you want, but I hope you'll consider it reasonable to allow the API to have three new items: `load`, `bypass`, and `loadPaths`

Otherwise, many (myself included) will end up replicating helper logic to convert an array of real locations for each package we use this in.

> The `bypass`, `enable`, and `disable` functions feel like they could be boiled down to a single function that works with either an async/promise returning function or a sync function. In the docs below, you warn against using async calls with the mock disabled - that same warning could be included in the `bypass` docs without adding the `enable` and `disable` functions.

Agreed. The consideration was that many people will simply use the function without noticing the warning. It was set up to be more strict to prevent people using it and filing issues. 

But this is entirely your call, if you're good with it, then it's no problem. I'll add async support, and we can add the warning to both the readme and JSDoc for the types package to make sure it's seen.

> Having a single `bypass` function also gets around the issue that calling `enable` twice makes `disable` not work (this is fixable, but could also be avoided by not adding these functions at all).

Didn't see that. I was up quite late last night. I'll correct the issue. If you're alright with it, I'd like to leave them attached to exports and simply not document in readme or in the @types package. That way there is zero-footprint, but people who know the source can use them. 

I can always replicate the behaviour to do it, but this way if anything changes my code doesn't break.

Let me know! Hopefully we can get this wrapped up. I'll wait until your responses before updating anything.

Reply:
I don't feel tediousness is an issue. I cannot imagine people use mock-fs to load lots of real dirs, it seems defeating the purpose of mocking.

Reply:
> I cannot imagine people use mock-fs to load lots of real dirs, it seems defeating the purpose of mocking.

Not at all. If we have multiple assets directories or even hard-coded paths on the system, this allows `mock-fs` to be used in a way that lets us 'mount' those areas and do whatever we like without actually modifying them. This is tremendously useful with a broad range of use-cases. 

That is, in fact, the reason I started this PR, but that's really all I can say on that. I don't want to belabor it further.

Reply:
> without actually modifying them

That's a good use case.